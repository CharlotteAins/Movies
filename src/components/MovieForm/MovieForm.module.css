.buttonsWrapper {
  display: flex;
  justify-content: flex-end;
  padding: 60px 0 20px;
}

.buttonBalancer {
  display: flex;
  justify-content: space-between;
  min-width: 450px;
}

.calendarIcon {
  position: absolute;
}

.calendarIcon:after {
  content: "";
  position: relative;
  left: 550px;
  top: 10px;
  display: block;
  width: 30px;
  height: 30px;
  color: #F65261;
  background-size: 30px 30px;
  background-image: url("/images/calendar-icon.svg");
  background-repeat: no-repeat;
  z-index: 5;
}
