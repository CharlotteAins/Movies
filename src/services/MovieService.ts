export interface Movie {
    id: number,
    title: string,
    imageUrl: string,
    tagline: string,
    releaseDate: Date
}

let mockMovies: Movie[] = [
    {id: 337167, title:'Fifty Shades Freed', imageUrl: 'https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg', releaseDate: new Date('2018-02-07'), tagline: 'Don\'t miss the climax'},
    {id: 338970, title:'Tomb Raider', imageUrl: 'https://image.tmdb.org/t/p/w500/ePyN2nX9t8SOl70eRW47Q29zUFO.jpg', releaseDate: new Date('2018-03-08'), tagline: 'Her legend begins'},
    {id: 181808, title:'Star Wars: The Last Jedi', imageUrl: 'https://image.tmdb.org/t/p/w500/kOVEVeg59E0wsnXmF9nrh6OmWII.jpg', releaseDate: new Date('2017-12-13'), tagline: 'The Saga Continues'},
    {id: 284054, title:'Black Panther', imageUrl: 'https://image.tmdb.org/t/p/w500/uxzzxijgPIY7slzFvMotPv8wjKA.jpg', releaseDate: new Date('2018-02-13'), tagline: 'Long live the king'},
    {id: 354912, title:'Coco', imageUrl: 'https://image.tmdb.org/t/p/w500/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg', releaseDate: new Date('2017-10-27'), tagline: 'The celebration of a lifetime'},
    {id: 333339, title:'Ready Player One', imageUrl: 'https://image.tmdb.org/t/p/w500/pU1ULUq8D3iRxl1fdX2lZIzdHuI.jpg', releaseDate: new Date('2018-03-28'), tagline: 'A better reality awaits.'},
]

export const MovieService = {

    getAllMovies: () => {
        return mockMovies;
    },

    editMovie: (id: number) => {
        console.log(`editing movie with id ${id} ...`)
    },

    deleteMovie: (id: number) => {
        console.log(`deleted movie with id ${id}`)
    }
}
