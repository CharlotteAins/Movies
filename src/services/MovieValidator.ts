export interface ValidationError {
  title?: string,
  overview?: string,
  runtime?: string,
  release_date?: string,
  poster_path?: string,
  genres?: string,
}


export const MovieValidator = ( values ) => {
  const errors: ValidationError = {};

  if ( !values.title.trim() ) {
    errors.title = 'Title is required';
  }

  if ( !values.overview.trim() ) {
    errors.overview = 'Overview is required';
  }

  if ( isNaN( values.runtime ) ) {
    errors.runtime = 'Runtime should be number';
  } else if ( values.runtime < 0 ) {
    errors.runtime = 'Runtime should be positive';
  }

  if ( !values.poster_path.trim() ) {
    errors.poster_path = 'Image URL is required';
  } else if ( !values.poster_path.match( /(http(s?):)([/|.|\w|\s|-])*\.(?:jpg|gif|png)/gi ) ) {
    errors.poster_path = 'Image URL should be correct';
  }

  if ( !values.genres.length ) {
    errors.genres = 'Genre should be at least one';
  }

  return errors;
};
