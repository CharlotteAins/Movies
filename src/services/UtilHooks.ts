import {useCallback, useState} from "react";

export const useToggle = (initialValue: boolean = false): any => {
    const [flag, setFlag] = useState(initialValue);

    const toggle = useCallback((value?: boolean) => {
        if (value) {
            setFlag(value)
        } else {
            setFlag(!flag);
        }
    }, [flag]);

    return [flag, toggle];
}
