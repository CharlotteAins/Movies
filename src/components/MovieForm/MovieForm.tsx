import React from 'react';
import {Movie} from "../../services/MovieService";
import Field from "../Field";
import styles from './MovieForm.module.css';
import PrettyButton from "../PrettyButton";
import CategorySelector from "../CategorySelector";
import {MovieValidator} from "../../services/MovieValidator";
import {useFormik} from 'formik';
import DateInput from "../DateInput";

interface MovieFormProps {
    formType: string,
    initialMovie: Movie,
    submitHandler: (movie: Movie) => void
}

const MovieForm: React.FC<MovieFormProps> = ({formType, initialMovie, submitHandler}) => {

    const formik = useFormik({
        initialValues: initialMovie,
        validate: MovieValidator,
        onSubmit: submitHandler
    });

    return (
        <form onSubmit={formik.handleSubmit}>
            <h3>{formType} movie</h3>

            {formType === 'edit' &&
            <>
                <label>movie id</label>
                <div>{formik.values.id}</div>
            </>}

            <Field
                inputKey='title'
                label='title'
                type='text'
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
                value={formik.values.title}
                placeholder='enter title'
                error={formik.errors.title}/>

            <Field
                inputKey='overview'
                label='overview'
                type='text'
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
                value={formik.values.overview}
                placeholder='enter overview'
                error={formik.errors.overview}/>

            <Field
                inputKey='runtime'
                label='runtime'
                type='number'
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
                value={formik.values.runtime}
                placeholder='runtime text goes here'
                error={formik.errors.runtime}/>

            <DateInput
                startDate={formik.values.release_date}
                dateHandler={formik.setFieldValue}
            />

            <Field
                inputKey='poster_path'
                label='movie url'
                type='text'
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
                value={formik.values.poster_path}
                placeholder='movie URL here'
                error={formik.errors.poster_path}/>

            <CategorySelector
                selectedCategories={formik.values.genres}
                chooseCategoryHandler={formik.setFieldValue}
            />

            <div className={styles.buttonsWrapper}>
                <div className={styles.buttonBalancer}>
                    <PrettyButton clickHandler={() => formik.resetForm()}
                                  text='reset'/>
                    <PrettyButton clickHandler={() => formik.handleSubmit()}
                                  text={formType === 'add' ? 'save' : 'submit'}/>
                </div>
            </div>
        </form>
    );

};

export default MovieForm;
