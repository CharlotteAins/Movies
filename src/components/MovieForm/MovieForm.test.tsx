import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import MovieForm from './MovieForm';
import {Movie} from '../../services/MovieService';

describe('MovieForm', () => {
    const mockHandler = jest.fn();
    let emptyMovie: Movie;

    const renderMovieForm = () => render(<MovieForm formType={'add'} initialMovie={emptyMovie} submitHandler={mockHandler}/>);

    beforeEach(() => {
        emptyMovie = {release_date: null, poster_path: '', title: '', overview: '', runtime: 0, genres: []};
    });

    it('should render properly', () => {
        const tree = renderMovieForm();
        expect(tree).toMatchSnapshot();
    });

    it('should not call submitHandler if inputs empty', () => {
        renderMovieForm();

        fireEvent.click(screen.getByText('save'));

        expect(mockHandler).not.toBeCalled();
    });

    it('should work reset form', () => {
        renderMovieForm();
        const input: HTMLInputElement = screen.getByPlaceholderText('enter title') as HTMLInputElement;

        fireEvent.change(input, {target: {value: 'mock title'}});
        expect(input.value).toEqual('mock title');

        fireEvent.click(screen.getByText('reset'));
        expect(input.value).toEqual('');
    });

    it('should send correct form', () => {
        renderMovieForm();

        fireEvent.change(screen.getByPlaceholderText('enter title'), {target: {value: 'mock title'}});
        fireEvent.change(screen.getByPlaceholderText('enter overview'), {target: {value: 'mock overview'}});
        fireEvent.change(screen.getByPlaceholderText('runtime text goes here'), {target: {value: 120}});
        fireEvent.change(screen.getByPlaceholderText('Select Date'), {target: {value: new Date('2020-10-10T03:24:00')}});
        fireEvent.change(screen.getByPlaceholderText('movie URL here'), {target: {value: 'https://cdn.jpegmini.com/user/images/slider_puffin_before_mobile.jpg'}});
        fireEvent.click(screen.getByDisplayValue('action'));

        fireEvent.click(screen.getByText('save'));
        expect(mockHandler).toBeCalled();
    });
});
