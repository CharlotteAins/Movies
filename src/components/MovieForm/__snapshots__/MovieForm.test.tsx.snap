// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm should render properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <h3>
          add
           movie
        </h3>
        <label
          form="title"
        >
          title
        </label>
        <input
          id="title"
          placeholder="enter title"
          type="text"
          value=""
        />
        <span />
        <label
          form="overview"
        >
          overview
        </label>
        <input
          id="overview"
          placeholder="enter overview"
          type="text"
          value=""
        />
        <span />
        <label
          form="runtime"
        >
          runtime
        </label>
        <input
          id="runtime"
          placeholder="runtime text goes here"
          type="number"
          value="0"
        />
        <span />
        <label
          form="release_date"
        >
          realize date
        </label>
        <div
          class="calendarIcon"
        />
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              class=""
              id="release_date"
              name="release_date"
              placeholder="Select Date"
              type="text"
              value=""
            />
          </div>
        </div>
        <label
          form="poster_path"
        >
          movie url
        </label>
        <input
          id="poster_path"
          placeholder="movie URL here"
          type="text"
          value=""
        />
        <span />
        <label
          form="genre"
        >
          genre
        </label>
        <div
          class="inputWrapper"
        >
          <div
            class="selectorIcon"
          />
          <input
            disabled=""
            id="genre"
            placeholder="select genre"
            value=""
          />
        </div>
        <ul
          class="categoryList"
          hidden=""
        >
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                action
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="action"
                type="checkbox"
                value="action"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                adventure
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="adventure"
                type="checkbox"
                value="adventure"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                animation
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="animation"
                type="checkbox"
                value="animation"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                comedy
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="comedy"
                type="checkbox"
                value="comedy"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                crime
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="crime"
                type="checkbox"
                value="crime"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                documentary
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="documentary"
                type="checkbox"
                value="documentary"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                drama
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="drama"
                type="checkbox"
                value="drama"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                family
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="family"
                type="checkbox"
                value="family"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                fantasy
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="fantasy"
                type="checkbox"
                value="fantasy"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                horror
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="horror"
                type="checkbox"
                value="horror"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                romance
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="romance"
                type="checkbox"
                value="romance"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
          <li
            class="categoryItem"
          >
            <label
              class="container"
            >
              <span>
                science fiction
              </span>
              <input
                class="checkbox"
                id="genres"
                placeholder="science fiction"
                type="checkbox"
                value="science fiction"
              />
              <span
                class="checkmark"
              />
            </label>
          </li>
        </ul>
        <div
          class="buttonsWrapper"
        >
          <div
            class="buttonBalancer"
          >
            <button
              class="btnPretty"
              type="button"
            >
              reset
            </button>
            <button
              class="btnPretty"
              type="button"
            >
              save
            </button>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <h3>
        add
         movie
      </h3>
      <label
        form="title"
      >
        title
      </label>
      <input
        id="title"
        placeholder="enter title"
        type="text"
        value=""
      />
      <span />
      <label
        form="overview"
      >
        overview
      </label>
      <input
        id="overview"
        placeholder="enter overview"
        type="text"
        value=""
      />
      <span />
      <label
        form="runtime"
      >
        runtime
      </label>
      <input
        id="runtime"
        placeholder="runtime text goes here"
        type="number"
        value="0"
      />
      <span />
      <label
        form="release_date"
      >
        realize date
      </label>
      <div
        class="calendarIcon"
      />
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            id="release_date"
            name="release_date"
            placeholder="Select Date"
            type="text"
            value=""
          />
        </div>
      </div>
      <label
        form="poster_path"
      >
        movie url
      </label>
      <input
        id="poster_path"
        placeholder="movie URL here"
        type="text"
        value=""
      />
      <span />
      <label
        form="genre"
      >
        genre
      </label>
      <div
        class="inputWrapper"
      >
        <div
          class="selectorIcon"
        />
        <input
          disabled=""
          id="genre"
          placeholder="select genre"
          value=""
        />
      </div>
      <ul
        class="categoryList"
        hidden=""
      >
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              action
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="action"
              type="checkbox"
              value="action"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              adventure
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="adventure"
              type="checkbox"
              value="adventure"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              animation
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="animation"
              type="checkbox"
              value="animation"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              comedy
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="comedy"
              type="checkbox"
              value="comedy"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              crime
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="crime"
              type="checkbox"
              value="crime"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              documentary
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="documentary"
              type="checkbox"
              value="documentary"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              drama
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="drama"
              type="checkbox"
              value="drama"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              family
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="family"
              type="checkbox"
              value="family"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              fantasy
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="fantasy"
              type="checkbox"
              value="fantasy"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              horror
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="horror"
              type="checkbox"
              value="horror"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              romance
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="romance"
              type="checkbox"
              value="romance"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
        <li
          class="categoryItem"
        >
          <label
            class="container"
          >
            <span>
              science fiction
            </span>
            <input
              class="checkbox"
              id="genres"
              placeholder="science fiction"
              type="checkbox"
              value="science fiction"
            />
            <span
              class="checkmark"
            />
          </label>
        </li>
      </ul>
      <div
        class="buttonsWrapper"
      >
        <div
          class="buttonBalancer"
        >
          <button
            class="btnPretty"
            type="button"
          >
            reset
          </button>
          <button
            class="btnPretty"
            type="button"
          >
            save
          </button>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
